src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -567,10 +567,49 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
 
     def varDecl(pos: Position, mods: Modifiers, tpt: Tree, name: TermName): ValDef = {
       val tpt1 = optArrayBrackets(tpt)
-      if (in.token == EQUALS && !mods.isParameter) skipTo(COMMA, SEMI)
+
+      /** Tries to detect final static literals syntactically and returns a constant type replacement */
+      def optConstantTpe(): Tree = {
+        in.nextToken()
+
+        def constantTpe(lit: Any): Tree =
+          try TypeTree(ConstantType(Constant(lit)))
+          finally in.nextToken()
+
+        if (mods.hasFlag(Flags.STATIC) && mods.isFinal) {
+          def lit(negate: Boolean): Tree =
+            if (in.lookaheadToken == SEMI)
+              in.token match {
+                case INTLIT if tpt.tpe == IntTpe => constantTpe(in.intVal(negate).toInt)
+                case LONGLIT if tpt.tpe == LongTpe => constantTpe(in.intVal(negate))
+                case FLOATLIT if tpt.tpe == FloatTpe => constantTpe(in.floatVal(negate).toFloat)
+                case DOUBLELIT if tpt.tpe == DoubleTpe => constantTpe(in.floatVal(negate))
+                case STRINGLIT =>
+                  tpt match {
+                    case Ident(TypeName("String")) => constantTpe(in.name.toString)
+                    case _ => tpt1
+                  }
+                case _ => tpt1
+              }
+            else tpt1
+
+          in.token match {
+            case MINUS =>
+              in.nextToken()
+              lit(negate = true)
+            case other => lit(negate = false)
+          }
+        } else tpt1
+      }
+
+      val tpt2: Tree =
+        if (in.token == EQUALS && !mods.isParameter)
+          try optConstantTpe() finally skipTo(COMMA, SEMI)
```

Comment:
somehow an elegant pattern, but it's also a bit confusing because it makes me thing there are exceptions involved. and it generates handlers in the bytecode.